source/required-deep.d.ts

```diff
@@ -0,0 +1,62 @@
+import {BuiltIns} from './internal';
+
+type ExcludeUndefined<T> = Exclude<T, undefined>;
+
+/**
+Create a type from another type with all keys and nested keys set to optional.
+
+Use-cases:
+- Merging a default settings/config object with another object, the second object would be a deep partial of the default object.
+- Mocking and testing complex entities, where populating an entire object with its keys would be redundant in terms of the mock or test.
+
+@example
+```
+import type {RequiredDeep} from 'type-fest';
+
+type Settings = {
+	textEditor?: {
+		fontSize?: number | undefined;
+		fontColor?: string | undefined;
+		fontWeight?: number | undefined;
+	}
+	autocomplete?: boolean | undefined;
+	autosave?: boolean | undefined;
+};
+
+type RequiredSettings = RequiredDeep<Settings>;
+Result: {
+	textEditor: {
+		fontSize: number;
+		fontColor: string;
+		fontWeight: number;
+	}
+	autocomplete: boolean;
+	autosave: boolean;
+}
+```
+
+ @category Utilities
+ */
+export type RequiredDeep<T, E extends ExcludeUndefined<T> = ExcludeUndefined<T>> = E extends BuiltIns
+	? E
+	: E extends Map<infer KeyType, infer ValueType>
+		? Map<RequiredDeep<KeyType>, RequiredDeep<ValueType>>
+		: E extends Set<infer ItemType>
+			? Set<RequiredDeep<ItemType>>
+			: E extends ReadonlyMap<infer KeyType, infer ValueType>
+				? ReadonlyMap<RequiredDeep<KeyType>, RequiredDeep<ValueType>>
+				: E extends ReadonlySet<infer ItemType>
+					? ReadonlySet<RequiredDeep<ItemType>>
+					: E extends (arg: any[]) => unknown
```

Comment:
I think there is a small typo here, this type will only match  functions with a single array argument. From my (limited) testing it should actually be `: E extends (...arg: any[]) => unknown` to match all types of function args.

Reply:
@ilchenkoArtem Make sure you add a test for this too.

Reply:
I wil fix this it this week